### DISCUSSION TITLE
Custom Content Fields: add custom settings in the Content Field settings page

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15522] [Q&A] [answered]
Hi everyone,

I'm doing some tests on **custom content fields**, and I have created a module (feature) for my content fields, that contains a `MyText` field. I have also implemented a `MyTextFieldDisplayDriver` class that is a `ContentFieldDisplayDriver<MyTextField>`.

I noticed that when I add my custom content field to a _Content Type_, I can **click on the edit button**, as shown in this screenshot:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/0ce5a4a7-999c-483e-b2d9-a7807278df4c)

And **a settings page is displayed** (URL path: https://localhost:7825/Admin/ContentParts/Index/Fields/MyText/Edit?returnUrl=%2FAdmin%2FContentTypes%2FEdit%2FIndex):

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/2cf085f7-dd67-4ca5-a7a0-653a8e6628ee)

But, when adding breakpoints in my `MyTextFieldDisplayDriver` class, none of the methods (Display, Edit, and UpdateAsync) are triggered. 
**The main question is**: how can I add custom settings to my custom content field?

**What I'm attempting to do is** add a string property to my custom field, this string property will contains a regex. This will serve as a validation method when the Admin user fills out the field and submits it to the `UpdateAsync` method in my `MyTextFieldDisplayDriver`.

Of course, I know that I can also add a regex validation using a _custom editor_ (we already discussed this topic [here](https://github.com/OrchardCMS/OrchardCore/discussions/15417#discussioncomment-8686457)), but my goal is to explore whether I can achieve the same outcome with a custom validation attribute. I'm treating this as an experiment.

Thank you for your help 😊

**Comment by MarGraz** [comment]
I checked the source code and found the "Edit page" in the backoffice, specifically the [`EditPage.cshtml`](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/Admin/EditField.cshtml) and [`EditFieldViewModel`](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.ContentTypes/ViewModels/EditFieldViewModel.cs). It seems that is not possible to customize this page. Is that correct?

**Comment by MarGraz** [comment]
@hishamco apologies for the direct mention. I noticed that you contributed to the Content Types documentation [page](https://docs.orchardcore.net/en/main/docs/reference/modules/ContentTypes/), and I thought you might have a suggestion regarding the question I asked above. Do you have any insights? 🙂

Thank you

**Reply by hishamco** [reply]
No problem feel free to mention me when you need :)

Please share you minimal repo, so I can have a look

**Comment by douwinga** [comment, marked as answer]
You would need a settings driver for that. Which is different from the Field display driver. You will have a class that inherits from `ContentPartFieldDefinitionDisplayDriver`. Something like `public class MyTextFieldSettingsDriver : ContentPartFieldDefinitionDisplayDriver<MyTextField>` which is then added to your `Startup.cs` with something like `services.AddScoped<IContentPartFieldDefinitionDisplayDriver, MyTextFieldSettingsDriver>();`. You will have to add your view and such too, but this should get you in the right direction and give you something to search for.

The `ContentFieldDisplayDriver` is for handling the displaying on the front end and editing of the content item itself, not the editing of the settings.

**Reply by MarGraz** [reply]
Thank you for your reply.

Do you know if it’s possible to create a `SettingsDriver` class for the `Text` field class that is already defined in Orchard Core?

Thank you

**Reply by douwinga** [reply]
Yes you can. We do it. It is the same as with a custom field as described above, but replace `<MyTextField>` with `<TextField>`
